### ISSUE TITLE
Can't use validator with Rails 3.2 and DataMapper 1.2

### ISSUE THREAD

**Issue by joaomm** [krobertson/dm-paperclip#24]
I'm trying to use dm-validators:

config.use_dm_validations = true

But, I end up with the following error:

$path/gems/dm-core-1.2.0/lib/dm-core/model/relationship.rb:372:in `method missing': undefined method`opts_from_validator_args' for MyClass:Class (NoMethodError).

This is my gemfile:

```
source 'https://rubygems.org'

RAILS_VERSION = '~> 3.2.3'

#gem 'rails', '3.2.3'
gem 'activesupport', RAILS_VERSION, :require => 'active_support'
gem 'actionpack', RAILS_VERSION, :require => 'action_pack'
gem 'actionmailer', RAILS_VERSION, :require => 'action_mailer'
gem 'activeresource', RAILS_VERSION, :require => 'active_resource'
gem 'railties', RAILS_VERSION, :require => 'rails'
gem 'tzinfo', '~> 0.3.32'

gem 'dm-rails', '~> 1.2.1'
gem 'dm-sqlite-adapter', group: [:development, :test]
gem 'dm-postgres-adapter', group: [:production]
# You can use any of the other available database adapters.
# This is only a small excerpt of the list of all available adapters
# Have a look at
#
# http://wiki.github.com/datamapper/dm-core/adapters
# http://wiki.github.com/datamapper/dm-core/community-plugins
#
# for a rather complete list of available datamapper adapters and plugins
gem 'dm-migrations'
gem 'dm-types'
gem 'dm-validations'
gem 'dm-constraints'
gem 'dm-transactions'
gem 'dm-aggregates'
gem 'dm-timestamps'
gem 'dm-observer'
gem 'dm-core'
gem 'dm-serializer'
gem 'dm-constraints'

gem 'dm-devise',       '~> 2.1.0'
gem 'data_mapper'
gem 'dm-paperclip', git: 'https://github.com/krobertson/dm-paperclip.git'

gem 'responders'

gem "haml-rails"

gem "heroku"
gem "omniauth-facebook"
gem 'aws-s3'

# FaceBook
gem 'koala'

# MailChimp
gem 'hominid', '~> 3.0'

gem 'jquery-rails'

gem 'friendly_id', "~> 4.0"

group :assets do
  gem 'sass-rails', "  ~> 3.1"
  gem 'coffee-rails', "~> 3.1"
  gem 'uglifier', ">= 1.0.3"
end

group :development do
  gem "taps"
end

group :test do
  # Pretty printed test output
  gem 'turn', :require => false
  gem 'ffaker'  
  gem 'rspec-rails'
  gem 'shoulda-matchers'
  gem 'cucumber-rails'
  gem 'capybara'
  gem 'launchy'
  gem 'guard-rspec'
  gem 'guard-livereload'
  gem 'growl'
  gem 'factory_girl'
  gem 'factory_girl_generator'
  gem 'simplecov', require: false
end

group :production do
  gem "unicorn"
  gem "pg"
  gem "execjs"
  gem 'newrelic_rpm'
  gem 'dalli'
end
```

And the Gemfile.lock:

```
GIT
  remote: https://github.com/krobertson/dm-paperclip.git
  revision: aabad04f2a65e8f628e7e838d9bd3e1c9c39914d
  specs:
    dm-paperclip (2.4.1)

GEM
  remote: https://rubygems.org/
  specs:
    actionmailer (3.2.3)
      actionpack (= 3.2.3)
      mail (~> 2.4.4)
    actionpack (3.2.3)
      activemodel (= 3.2.3)
      activesupport (= 3.2.3)
      builder (~> 3.0.0)
      erubis (~> 2.7.0)
      journey (~> 1.0.1)
      rack (~> 1.4.0)
      rack-cache (~> 1.2)
      rack-test (~> 0.6.1)
      sprockets (~> 2.1.2)
    activemodel (3.2.3)
      activesupport (= 3.2.3)
      builder (~> 3.0.0)
    activeresource (3.2.3)
      activemodel (= 3.2.3)
      activesupport (= 3.2.3)
    activesupport (3.2.3)
      i18n (~> 0.6)
      multi_json (~> 1.0)
    addressable (2.2.8)
    ansi (1.4.2)
    aws-s3 (0.6.3)
      builder
      mime-types
      xml-simple
    bcrypt-ruby (3.0.1)
    builder (3.0.0)
    capybara (1.1.2)
      mime-types (>= 1.16)
      nokogiri (>= 1.3.3)
      rack (>= 1.0.0)
      rack-test (>= 0.5.4)
      selenium-webdriver (~> 2.0)
      xpath (~> 0.1.4)
    childprocess (0.3.2)
      ffi (~> 1.0.6)
    coffee-rails (3.2.2)
      coffee-script (>= 2.2.0)
      railties (~> 3.2.0)
    coffee-script (2.2.0)
      coffee-script-source
      execjs
    coffee-script-source (1.3.3)
    cucumber (1.2.0)
      builder (>= 2.1.2)
      diff-lcs (>= 1.1.3)
      gherkin (~> 2.10.0)
      json (>= 1.4.6)
    cucumber-rails (1.3.0)
      capybara (>= 1.1.2)
      cucumber (>= 1.1.8)
      nokogiri (>= 1.5.0)
    dalli (2.0.5)
    data_mapper (1.2.0)
      dm-aggregates (~> 1.2.0)
      dm-constraints (~> 1.2.0)
      dm-core (~> 1.2.0)
      dm-migrations (~> 1.2.0)
      dm-serializer (~> 1.2.0)
      dm-timestamps (~> 1.2.0)
      dm-transactions (~> 1.2.0)
      dm-types (~> 1.2.0)
      dm-validations (~> 1.2.0)
    data_objects (0.10.8)
      addressable (~> 2.1)
    devise (2.1.0)
      bcrypt-ruby (~> 3.0)
      orm_adapter (~> 0.0.7)
      railties (~> 3.1)
      warden (~> 1.1.1)
    diff-lcs (1.1.3)
    dm-active_model (1.2.1)
      activemodel (~> 3.0)
      dm-core (~> 1.2.0)
    dm-aggregates (1.2.0)
      dm-core (~> 1.2.0)
    dm-constraints (1.2.0)
      dm-core (~> 1.2.0)
    dm-core (1.2.0)
      addressable (~> 2.2.6)
    dm-devise (2.1.0)
      devise (>= 1.3)
      dm-core (>= 1.0.2, < 2.0)
      dm-migrations (>= 1.0.2, < 2.0)
      dm-rails (>= 1.0.2, < 2.0)
      dm-serializer (>= 1.0.2, < 2.0)
      dm-timestamps (>= 1.0.2, < 2.0)
    dm-do-adapter (1.2.0)
      data_objects (~> 0.10.6)
      dm-core (~> 1.2.0)
    dm-migrations (1.2.0)
      dm-core (~> 1.2.0)
    dm-observer (1.2.0)
      dm-core (~> 1.2.0)
    dm-postgres-adapter (1.2.0)
      dm-do-adapter (~> 1.2.0)
      do_postgres (~> 0.10.6)
    dm-rails (1.2.1)
      actionpack (~> 3.0)
      dm-active_model (~> 1.2.0)
      dm-core (~> 1.2.0)
      railties (~> 3.0)
    dm-serializer (1.2.1)
      dm-core (~> 1.2.0)
      fastercsv (~> 1.5.4)
      json (~> 1.6.1)
      json_pure (~> 1.6.1)
      multi_json (~> 1.0.3)
    dm-sqlite-adapter (1.2.0)
      dm-do-adapter (~> 1.2.0)
      do_sqlite3 (~> 0.10.6)
    dm-timestamps (1.2.0)
      dm-core (~> 1.2.0)
    dm-transactions (1.2.0)
      dm-core (~> 1.2.0)
    dm-types (1.2.1)
      bcrypt-ruby (~> 3.0.0)
      dm-core (~> 1.2.0)
      fastercsv (~> 1.5.4)
      json (~> 1.6.1)
      multi_json (~> 1.0.3)
      stringex (~> 1.3.0)
      uuidtools (~> 2.1.2)
    dm-validations (1.2.0)
      dm-core (~> 1.2.0)
    do_postgres (0.10.8)
      data_objects (= 0.10.8)
    do_sqlite3 (0.10.8)
      data_objects (= 0.10.8)
    em-websocket (0.3.6)
      addressable (>= 2.1.1)
      eventmachine (>= 0.12.9)
    erubis (2.7.0)
    eventmachine (0.12.10)
    excon (0.13.4)
    execjs (1.4.0)
      multi_json (~> 1.0)
    factory_girl (3.3.0)
      activesupport (>= 3.0.0)
    factory_girl_generator (0.0.3)
    faraday (0.8.1)
      multipart-post (~> 1.1)
    fastercsv (1.5.5)
    ffaker (1.14.0)
    ffi (1.0.11)
    friendly_id (4.0.6)
    gherkin (2.10.0)
      json (>= 1.4.6)
    growl (1.0.3)
    guard (1.0.3)
      ffi (>= 0.5.0)
      thor (>= 0.14.6)
    guard-livereload (0.4.2)
      em-websocket (>= 0.2.0)
      guard (>= 0.10.0)
      multi_json (~> 1.0)
    guard-rspec (0.7.3)
      guard (>= 0.10.0)
    haml (3.1.6)
    haml-rails (0.3.4)
      actionpack (~> 3.0)
      activesupport (~> 3.0)
      haml (~> 3.0)
      railties (~> 3.0)
    hashie (1.2.0)
    heroku (2.26.2)
      heroku-api (~> 0.2.1)
      launchy (>= 0.3.2)
      netrc (~> 0.7.2)
      rest-client (~> 1.6.1)
      rubyzip
    heroku-api (0.2.1)
      excon (~> 0.13.3)
    hike (1.2.1)
    hominid (3.0.4)
    httpauth (0.1)
    i18n (0.6.0)
    journey (1.0.3)
    jquery-rails (2.0.2)
      railties (>= 3.2.0, < 5.0)
      thor (~> 0.14)
    json (1.6.7)
    json_pure (1.6.7)
    kgio (2.7.4)
    koala (1.1.0)
      multi_json (~> 1.0)
      multipart-post (~> 1.0)
    launchy (2.1.0)
      addressable (~> 2.2.6)
    libwebsocket (0.1.3)
      addressable
    mail (2.4.4)
      i18n (>= 0.4.0)
      mime-types (~> 1.16)
      treetop (~> 1.4.8)
    mime-types (1.18)
    multi_json (1.0.4)
    multipart-post (1.1.5)
    netrc (0.7.2)
    newrelic_rpm (3.3.4.1)
    nokogiri (1.5.2)
    oauth2 (0.6.0)
      faraday (~> 0.7)
      httpauth (~> 0.1)
      multi_json (~> 1.0)
    omniauth (1.1.0)
      hashie (~> 1.2)
      rack
    omniauth-facebook (1.3.0)
      omniauth-oauth2 (~> 1.0.2)
    omniauth-oauth2 (1.0.2)
      oauth2 (~> 0.6.0)
      omniauth (~> 1.0)
    orm_adapter (0.0.7)
    pg (0.13.2)
    polyglot (0.3.3)
    rack (1.4.1)
    rack-cache (1.2)
      rack (>= 0.4)
    rack-ssl (1.3.2)
      rack
    rack-test (0.6.1)
      rack (>= 1.0)
    railties (3.2.3)
      actionpack (= 3.2.3)
      activesupport (= 3.2.3)
      rack-ssl (~> 1.3.2)
      rake (>= 0.8.7)
      rdoc (~> 3.4)
      thor (~> 0.14.6)
    raindrops (0.9.0)
    rake (0.9.2.2)
    rdoc (3.12)
      json (~> 1.4)
    responders (0.9.1)
      railties (~> 3.1)
    rest-client (1.6.7)
      mime-types (>= 1.16)
    rspec (2.10.0)
      rspec-core (~> 2.10.0)
      rspec-expectations (~> 2.10.0)
      rspec-mocks (~> 2.10.0)
    rspec-core (2.10.1)
    rspec-expectations (2.10.0)
      diff-lcs (~> 1.1.3)
    rspec-mocks (2.10.1)
    rspec-rails (2.10.1)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      railties (>= 3.0)
      rspec (~> 2.10.0)
    rubyzip (0.9.8)
    sass (3.1.19)
    sass-rails (3.2.5)
      railties (~> 3.2.0)
      sass (>= 3.1.10)
      tilt (~> 1.3)
    selenium-webdriver (2.22.0)
      childprocess (>= 0.2.5)
      ffi (~> 1.0)
      libwebsocket (~> 0.1.3)
      multi_json (~> 1.0)
      rubyzip
    sequel (3.20.0)
    shoulda-matchers (1.1.0)
      activesupport (>= 3.0.0)
    simplecov (0.6.4)
      multi_json (~> 1.0)
      simplecov-html (~> 0.5.3)
    simplecov-html (0.5.3)
    sinatra (1.0)
      rack (>= 1.0)
    sprockets (2.1.3)
      hike (~> 1.2)
      rack (~> 1.0)
      tilt (~> 1.1, != 1.3.0)
    stringex (1.3.3)
    taps (0.3.24)
      rack (>= 1.0.1)
      rest-client (>= 1.4.0, < 1.7.0)
      sequel (~> 3.20.0)
      sinatra (~> 1.0.0)
    thor (0.14.6)
    tilt (1.3.3)
    treetop (1.4.10)
      polyglot
      polyglot (>= 0.3.1)
    turn (0.9.5)
      ansi
    tzinfo (0.3.33)
    uglifier (1.2.4)
      execjs (>= 0.3.0)
      multi_json (>= 1.0.2)
    unicorn (4.3.1)
      kgio (~> 2.6)
      rack
      raindrops (~> 0.7)
    uuidtools (2.1.2)
    warden (1.1.1)
      rack (>= 1.0)
    xml-simple (1.1.1)
    xpath (0.1.4)
      nokogiri (~> 1.3)

PLATFORMS
  ruby

DEPENDENCIES
  actionmailer (~> 3.2.3)
  actionpack (~> 3.2.3)
  activeresource (~> 3.2.3)
  activesupport (~> 3.2.3)
  aws-s3
  capybara
  coffee-rails (~> 3.1)
  cucumber-rails
  dalli
  data_mapper
  dm-aggregates
  dm-constraints
  dm-core
  dm-devise (~> 2.1.0)
  dm-migrations
  dm-observer
  dm-paperclip!
  dm-postgres-adapter
  dm-rails (~> 1.2.1)
  dm-serializer
  dm-sqlite-adapter
  dm-timestamps
  dm-transactions
  dm-types
  dm-validations
  execjs
  factory_girl
  factory_girl_generator
  ffaker
  friendly_id (~> 4.0)
  growl
  guard-livereload
  guard-rspec
  haml-rails
  heroku
  hominid (~> 3.0)
  jquery-rails
  koala
  launchy
  newrelic_rpm
  omniauth-facebook
  pg
  railties (~> 3.2.3)
  responders
  rspec-rails
  sass-rails (~> 3.1)
  shoulda-matchers
  simplecov
  taps
  turn
  tzinfo (~> 0.3.32)
  uglifier (>= 1.0.3)
  unicorn
```

**Comment by kot-begemot** [comment]
I manage to solve this issue... how ever the bugfix suppose to be used carefully... 
I changed there `flush_writes` method, that was responsible for relocating temp files...
Unfortunately, for now, something like 

```
has_attached_file :resource,
  storage: :filesystem,
  path: "public/media/:basename_:style.:extension"
```

Will not work. However

```
has_attached_file :resource,
  storage: :filesystem,
  path: "public/media/:style/:basename.:extension"
```

That should work as expected. If you will run into any exceptions please let me know, will try to fix it.

**Comment by krobertson** [comment]
Fix in with #30
